OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md

```diff
@@ -0,0 +1,288 @@
+        200:
+          description: Success!
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/github.release"
+        404:
+          description: Not found
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/github.error"
+
+components:
+  schemas:
+    github.release:
+      type: object
+      properties:
+        id:
+          type: integer
+          description: The release ID
+        tag_name:
+          type: string
+          description: The release tag name
+        name:
+          type: string
+          description: The release name
+        body:
+          type: string
+          description: The release body
+        created_at:
+          type: string
+          format: date-time
+          description: The release creation date
+        published_at:
+          type: string
+          format: date-time
+          description: The release publication date
+    github.error:
+      type: object
+      properties:
+        message:
+          type: string
+          description: The error message
+        documentation_url:
+          type: string
+          description: The error documentation URL
+```
+
+You can then use the [Kiota](/openapi/kiota/install?tabs=bash&wt.mc_id=SEC-MVP-5004985#install-as-net-tool) command line tool to generate the API client classes. We are using the [deserializer](/openapi/kiota/using#--deserializer---ds) and [serializer](/openapi/kiota/using#--serializer---s) options to specify the serialization format, to only use JSON serialization.
+
+```bash
+kiota generate -l csharp -d github-releases.yml -c GitHubClient -n GitHub.ApiClient -o ./GitHub --deserializer Microsoft.Kiota.Serialization.Json.JsonParseNodeFactory --serializer Microsoft.Kiota.Serialization.Json.JsonSerializationWriterFactory
+```
+
+## Create extension methods
+
+Create a new file named _KiotaServiceCollectionExtensions.cs_ in the root of your project and add the following code.
+
+```csharp
+using Microsoft.Kiota.Http.HttpClientLibrary.Middleware;
+using Microsoft.Kiota.Http.HttpClientLibrary.Middleware.Options;
+
+/// <summary>
+/// Service collection extensions for Kiota handlers.
+/// </summary>
+public static class KiotaServiceCollectionExtensions
+{
+    /// <summary>
+    /// Adds the Kiota handlers to the service collection.
+    /// </summary>
+    /// <param name="services"><see cref="IServiceCollection"/> to add the services to</param>
+    /// <returns><see cref="IServiceCollection"/> as per convention</returns>
+    /// <remarks>The handlers are added to the http client by the <see cref="AttachKiotaHandlers(IHttpClientBuilder)"/> call, which requires them to be pre-registered in DI</remarks>
+    public static IServiceCollection AddKiotaHandlers(this IServiceCollection services)
+    {
+        services.AddTransient<UriReplacementHandler<UriReplacementHandlerOption>>();
```

Comment:
out of curiosity here, why not simply call that and feed the array in the services?
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L50

Reply:
This call is initializing them, and I just need the the types to register them in DI.
Or would you say that doing a 

```csharp

var handlers = KiotaClientFactory.CreateDefaultHandlers()
foreach(var handler in handlers) {
  services.AddTransient(typeof(handler));
}
```

is better? Just trying to see what looks better in coding structures that are commonly used.

Reply:
The challenge with explicitly enumerating the type is that if tomorrow we add a the middleware handler (let's say something that polls automatically long running operations for example) to the stack, people wouldn't get it automatically.
That's why if we can, it'd be better if it could be more dynamic on that part, even if it means a little more complex code.
does that make sense?

Reply:
I've made them dynamically, but left the old code. I really don't like this new way because it looks really ugly. And it's creating those handlers during startup twice, just to get the type of them.

Reply:
assuming this gets implemented, we should be able to clean it up here.
https://github.com/microsoft/kiota-http-dotnet/issues/254

Reply:
I just noticed that the new way does not work because of this one.
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L55

Reply:
I believe this was resolved with the update of http. Resolving